keel-sql/src/main/resources/db/changelog/20191108-environment-artifact-constraint.yml

```diff
@@ -93,3 +93,73 @@ databaseChangeLog:
             oldColumnName: json
             newColumnName: attributes
             columnDataType: longtext
+
+  # work around for https://liquibase.jira.com/browse/CORE-2111 to get the new uid column
+  # in the first position, since it's going to be the primary key
+  - changeSet:
+      id: mysql-envconstraint-add-uid
+      author: asher
+      changes:
+        - sql:
+            dbms: mysql
+            sql: alter table `environment_artifact_constraint` add column `uid` char(26) not null first
+
+  - changeSet:
+      id: envconstraint-add-uid
+      author: asher
+      changes:
+        - dropPrimaryKey:
+            tableName: environment_artifact_constraint
+            constraintName: constraint_pk
+            dropIndex: true
+        - renameColumn:
+            tableName: environment_artifact_constraint
+            columnDataType: varchar(255)
+            oldColumnName: constraint_type
+            newColumnName: type
+        - addPrimaryKey:
+            tableName: environment_artifact_constraint
+            constraintName: constraint_pk
+            columnNames: uid
+        - addUniqueConstraint:
+            tableName: environment_artifact_constraint
+            constraintName: constraint_pk
+            columnNames: environment_uid, type, artifact_version
+
+  - changeSet:
+      id: create-current-constraint-table
+      author: asher
+      changes:
+        - createTable:
+            tableName: current_constraint
+            columns:
+              - column:
+                  name: application
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: environment_uid
+                  type: char(26)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: type
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: constraint_uid
+                  type: char(26)
+                  constraints:
+                    nullable: false
+        - addPrimaryKey:
+            tableName: current_constraint
+            constraintName: current_constraint_pk
+            columnNames: application, environment_uid, type
+        - createIndex:
+            tableName: current_constraint
+            indexName: current_constraint_uid_idx
+            columns:
+              - column:
+                  name: constraint_uid
```

Comment:
Shouldn't these changes go into a new YAML file?

Reply:
Why?

Reply:
I'm not familiar with the framework we're using for DB migrations, but I was just commenting based on prior experience with other such tools where a "golden rule" of migrations was that, once committed into source control, you never changed the original migration script since different developers might have different versions of the code checked out. It's not a big deal in this case since you're just adding to the file, but my question was, again, more around the patterns we're following for my own education.

Reply:
We should consider each `changeSet` within a migration file to be immutable (liquibase tracks the md5 hash of individual changeSets, not files) but don't have rules established beyond that. It would be nice to collapse the schema for public consumption when there's lots of refactoring between beginning and releasing a feature but we historically haven't done that solely due to it complicating the release process within netflix. When modifying a table that is already in use and defined by one or more existing liquibase files, I tend to create a new file but I think there's room for judgement there too.